### PR TITLE
Add profile service tests & fix profile service

### LINKED ISSUE #103: Tests for profile service

### PR DESCRIPTION
**Pull request by akulsharma1**
Resolves #103 

- Added `profile-router.test.ts` 
- Bug fixes to `profile-router.ts`

### CODE REVIEW

**Comment by Timothy-Gonzalez** [comment]
Several notes (not sure if significant, cc @AydanPirani for final review):

[Coverage report](https://github.com/HackIllinois/adonix/actions/runs/6580703800/job/17879249657?pr=102) for profile is as follows:
```
src/services/profile     |    83.8 |    73.91 |      25 |    83.8 |      
  profile-formats.ts     |   77.77 |       25 |     100 |   77.77 | 11-12,15-16,24-25
  profile-lib.ts         |   55.81 |      100 |       0 |   55.81 | 12-18,25-27,35-43
  profile-router.ts      |   88.65 |    84.21 |     100 |   88.65 | 53-77,166-167,228-229,244-246
```

Not covered by tests:
- profile-router.ts
  - 55-77: `GET /leaderboard/` - no idea how this works or if this should be tested yet
  - 166-167: Redirect condition when no user id, prob not easy to test but should consider checking
  - 228-229: Validation check for profile, should test this by sending bad data
  - 244-246: Check for error handling of failed create, probably not needed to test
- profile-formats.ts
  - We could add tests for this, but shouldn't as we should use something like zod instead, re: #104
- profile-lib.ts
  - 12-18: castLeaderboardEntries: used in `GET /leaderboard/`, will be covered by profile router tests
  - 25-27: isValidLimit: used in `GET /leaderboard/`, will be covered by profile router tests
  - 35-43: updatePoints: not used at all currently

All in all, pretty good coverage!

Some things I'd like to see tested:
- [ ] Post as a user (unauthorized user doesn't need to be tested, probably switch this out)
  - Basically, regular users can't edit their profile
- [ ] Can probably remove unauthorized user tests (called invalid user) as these will be tested in auth middleware

Other misc things / nitpicks:
- [ ] Can we rename "invalid user" to something more specific
  - for example: `it("delete with a user that doesn't exist")` -> `it("fails to delete with a non-attendee user")`
  - Try to make it a sentence you can read basically

This is a great start though, great job!

**Comment by AydanPirani** [comment]
@Timothy-Gonzalez  @akulsharma1 


### Leaderboard 
> GET /leaderboard/ - no idea how this works or if this should be tested yet

> 12-18: castLeaderboardEntries: used in GET /leaderboard/, will be covered by profile router tests

> 25-27: isValidLimit: used in GET /leaderboard/, will be covered by profile router tests

I wrote this super basic endpoint, it should return the top limit users by points - @akulsharma1  do you think you can add some tests in for this as well?  Basically, all you need to do is:

- Add in like 3-4 users with different point values
- Check that the endpoint (no limit) returns the users in ordered point values (most to least).
- Check that the endpoint (limit 2) returns the 2 users with the highest point counts.

And maybe add in a test case to ensure that limit cannot be negative.


### Format Testing
> We could add tests for this, but shouldn't as we should use something like zod instead, re: https://github.com/HackIllinois/adonix/issues/104

Yeah, dw about testing the actual "isValidformat" for now, just make sure that it throws if it gets a bad input.

### Role Stuff 

> Can probably remove unauthorized user tests (called invalid user) as these will be tested in auth middleware

Yup, dw about testing with roles that don't have any perms - just test that each branch of hasElevatedPerms/hasStaffPerms/hasAdminPerms gets tested.

**Review by AydanPirani — COMMENTED** [review]
it's making me review, but please reference above comments! I'm good with this once you get through the checkboxes + if @Timothy-Gonzalez  is happy. with it

File: src/services/profile/profile-router.test.ts [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+                discordTag: "test",
+            })
+            .expect(400);
+
+        expect(JSON.parse(response.text)).toHaveProperty("error", "InvalidParams");
+    });
+});
+
+describe("GET /profile", () => {
+    it("gets with an attendee that doesn't exist", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets with an attendee that doesn't exist")` -> `it("fails to get a profile that doesn't exist")`

File: src/services/profile/profile-router.test.ts [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        const response = await getAsUser("/profile/").expect(404);
+
+        expect(JSON.parse(response.text)).toHaveProperty("error", "UserNotFound");
+    });
+
+    it("gets an attendee that exists", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets an attendee that exist")` -> `it("gets a profile that exists")`

File: src/services/profile/profile-router.test.ts [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            points: 0,
+        });
+
+        const response = await getAsUser("/profile/").expect(200);
+        expect(JSON.parse(response.text)).toHaveProperty("displayName", TESTER.name);
+    });
+});
+
+describe("GET /profile/id/:USERID", () => {
+    it("gets without a :USERID", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets without a :USERID")` -> `it("fails with no id provided")`

File: src/services/profile/profile-router.test.ts [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        await Models.AttendeeProfile.create({
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: General overall thing: you're duplicating this same creation of an AttendeeProfile, make this attendee a variable (see [my pr](https://github.com/HackIllinois/adonix/pull/105) if you aren't sure the proper way to do this), also could probably use a beforeEach (also see [my pr](https://github.com/HackIllinois/adonix/pull/105))

File: src/services/profile/profile-router.test.ts [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            displayName: TESTER.name,
+            avatarUrl: TESTER.avatarUrl,
+            discordTag: TESTER.discordTag,
+            points: 0,
+        });
+
+        await getAsUser("/profile/id").expect(302);
+    });
+
+    it("gets with a user", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets with a user")` -> `it("fails to get a profile as a user")`

File: src/services/profile/profile-router.test.ts [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+            discordTag: TESTER.discordTag,
+            points: 0,
+        });
+
+        const response = await getAsUser("/profile/id/abcdefgabcdefgaaa").expect(403);
+
+        expect(JSON.parse(response.text)).toHaveProperty("error", "Forbidden");
+    });
+
+    it("gets with an admin", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets with an admin")` -> `it("gets a profile as an admin")`

File: src/services/profile/profile-router.test.ts [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            discordTag: TESTER.discordTag,
+            points: 0,
+        });
+
+        const response = await getAsAdmin("/profile/id/abcdefgabcdefgaaa").expect(200);
+
+        expect(JSON.parse(response.text)).toHaveProperty("displayName", "tst");
+    });
+
+    it("gets a user that doesnt exist", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts:118]: np: `it("gets a user that doesnt exist")` -> `it("fails to get a profile that doesn't exist")`

File: src/services/profile/profile-router.test.ts [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        });
+
+        const response = await getAsAdmin("/profile/id/doesnotexist").expect(404);
+
+        expect(JSON.parse(response.text)).toHaveProperty("error", "UserNotFound");
+    });
+});
+
+describe("DELETE /profile/", () => {
+    it("deletes with a attendee that doesn't exist", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("deletes with a attendee that doesn't exist")` -> `it("fails to delete a profile that doesn't exist")`

File: src/services/profile/profile-router.test.ts [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+        const response = await delAsUser("/profile").expect(404);
+        expect(JSON.parse(response.text)).toHaveProperty("error", "AttendeeNotFound");
+    });
+
+    it("deletes with a valid user", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("deletes with a valid user")` -> `it("deletes a profile")`

File: src/services/profile/profile-router.test.ts [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+            foodWave: 0,
+        });
+
+        const response = await delAsUser("/profile").expect(200);
+        expect(JSON.parse(response.text)).toHaveProperty("success", true);
+    });
+});
+
+describe("GET /profile/leaderboard", () => {
+    it("gets with no limit but only 3 entries", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets with no limit but only 3 entries")` -> `it("gets 3 entries when no limit is set")`

File: src/services/profile/profile-router.test.ts [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+            displayName: TESTER.name + "3",
+            avatarUrl: TESTER.avatarUrl,
+            discordTag: TESTER.discordTag,
+            points: 27,
+        });
+
+        await getAsUser("/profile/leaderboard?limit=2").expect(200);
+    });
+
+    it("gets with no limit & more entries in db than the max query limit", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets with no limit & more entries in db than the max query limit")` -> `it("only gets the max limit when no limit is set")`

File: src/services/profile/profile-router.test.ts [hunk 12]
```diff
@@ -0,0 +1,231 @@
...
+        }
+
+        const response = await getAsUser("/profile/leaderboard").expect(200);
+
+        const responseArray = JSON.parse(response.text);
+
+        expect(responseArray.profiles.length).toBeLessThan(Constants.LEADERBOARD_QUERY_LIMIT + 1);
+    });
+
+    it("gets with an invalid limit", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("gets with an invalid limit")` -> `it("fails when an invalid limit is set")`

**Review by Timothy-Gonzalez — CHANGES_REQUESTED** [review]
Looks great! A few small nps to move duplicated code out of tests & into variables & beforeEach hooks, but otherwise great job!

File: src/services/profile/profile-router.test.ts [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+const profile: ProfileFormat = {
+    userId: TESTER.id,
+    displayName: TESTER.name,
+    avatarUrl: TESTER.avatarUrl,
+    discordTag: TESTER.discordTag,
+    points: 0,
+};
+
+describe("POST /profile", () => {
+    it("posts for an attendee", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("posts for an attendee")` -> `it("works for an attendee")`

File: src/services/profile/profile-router.test.ts [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        const response = await postAsAttendee("/profile/").send(profile).expect(200);
+
+        expect(JSON.parse(response.text)).toHaveProperty("displayName", TESTER.name);
+    });
+
+    it("posts for a user", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("posts for a user")` -> `it("fails when a profile is already created")`

File: src/services/profile/profile-router.test.ts [hunk 15]
```diff
@@ -0,0 +1,231 @@
...
+        const response = await postAsUser("/profile/").send(profile).expect(200);
+
+        expect(JSON.parse(response.text)).toHaveProperty("displayName", TESTER.name);
+
+        // to verify they can't double create
+        const response2 = await postAsUser("/profile/").send(profile).expect(400);
+        expect(JSON.parse(response2.text)).toHaveProperty("error", "UserAlreadyExists");
+    });
+
+    it("posts with invalid data", async () => {
```
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts]: np: `it("posts with invalid data")` -> `it("fails when invalid data is provided")`

**Review by Timothy-Gonzalez — CHANGES_REQUESTED** [review]
few more nps that I missed

File: src/services/profile/profile-router.test.ts
Comment (Timothy-Gonzalez) [inline src/services/profile/profile-router.test.ts:1]: One final thing: we should use `StatusCode` enum introduced in #106 instead of constants like 200. Sorry I forgot to mention this before.

**Review by Timothy-Gonzalez — APPROVED** [review]
Great job! Feel free to merge!

*This should be a squash merge, make sure to edit the description to only include the important things (not "testing formatting adjustments" for example)*
